### ISSUE TITLE
realtime status API

### ISSUE THREAD

**Issue by meejah** [LeastAuthority/magic-folder#335]
To support GUI clients such as GridSync a status endpoint is desired. UI applications will connect via a WebSocket connection. In order that a UI doesn't miss any messages all messages will be buffered until at least one client connects to the status API; it will then receive all pending messages and begin receiving new ones in real time.

The status API shall report ongoing upload or download activity as a message indicating the total size of pending operations and their current percent completed. It is also desirable to have notification of new files added (that's "new LocalSnapshot added" I guess). Some notification of "idle" versus "synchronizing" is desired as well.

Currently the only concrete user is GridSync and it only currently has UI for showing a "sync-ing" indicator and the percent progress.

**Comment by meejah** [comment]
The code in 340 adds the general infrastructure for this feature as well as a single sort of message:

```
{
    "kind": "synchronizing",
    "status": true
}
```

The "status" will be `true` whenever there is work to do and `false` when there isn't. All messages are buffered whenever zero clients are connected (so e.g. GridSync should never miss any messages).

**Comment by meejah** [comment]
For lack of an immediate better place to put this, here is some example code that can use this API...

```python
import json

from autobahn.twisted.websocket import (
    WebSocketClientProtocol,
    WebSocketClientFactory,
)
from twisted.internet.endpoints import (
    clientFromString,
)
from twisted.internet.task import (
    react,
)
from twisted.internet.defer import (
    Deferred,
)


class StatusProtocol(WebSocketClientProtocol):
    """
    Client-side protocol that receives status updates from
    magic-folder
    """

    def onOpen(self):
        print("WebSocket connection open.")

    def onMessage(self, payload, isBinary):
        data = json.loads(payload)
        print("status: {kind}={status}".format(**data))

    def onClose(self, wasClean, code, reason):
        print("WebSocket connection closed: {0}".format(reason))


if __name__ == '__main__':

    def main(reactor):
        with open("alice/api_token", "r") as f:
            token = f.read().strip()
        factory = WebSocketClientFactory(
            "ws://127.0.0.1:4444/v1/status",
            headers={
                "Authorization": "Bearer {}".format(token),
            }
        )
        factory.protocol = StatusProtocol
        endpoint = clientFromString(reactor, "tcp:127.0.0.1:4444")
        endpoint.connect(factory)
        return Deferred()
    react(main)
```

**Comment by meejah** [comment]
From @tomprince in https://github.com/LeastAuthority/magic-folder/pull/340#discussion_r638330790 :

I'm not sure that it makes sense to keep track of messages when nobody is connected (particularly not an unbounded amount). Given that we don't persist a history, I think any client will need to get the current state independent of the websocket.

My impression of how anything subscribing to a status feed like this works is 1) connect to the status feed, 2) query current state using info from the feed to incrementally update it.

**Comment by meejah** [comment]
As I understand this from GridSync's perspective, it's going to start up `magic-folder` and then "pretty soon after that" connect to the status-API. However, it could be that magic-folder decided to queue some work to do and so GridSync could miss some status updates about that queued work.

Since the work-queues are ephemeral, it doesn't really make sense to persist the updates but it *does* make sense to keep them in memory so that GridSync (or whatever UI client) doesn't miss anything. I know that right now we're just emitting a couple pretty simple messages, but we do expect that to change.

We *could* of course add APIs for querying the state of the work-queues but I feel like that's only necessary if those work-queues themselves are persistent (because then, yes, UI clients would have to start up magic-folder and then query the state of its persistent work-queue).

The advantage of buffering the messages is that it's a very simple API: a UI client connects to it and is assured it has received all status updates sine the process was started. Since the "work to do" is discovered at start-up, this will be all the work. I guess on the "minus" side it only supports a single client properly (we could enforce that, though).

Otherwise we're building a whole "discover the state of stuff" API for state that's a) ephemeral anyway and b) probably only matters for "hundreds of milliseconds" at startup or if GridSync needs to re-connect for some reason.

**Comment by tomprince** [comment]
>  a UI client connects to it and is assured it has received all status updates sine the process was started.

This is only true if no other client has connected between when the service has started and when a given client connects. I don't think we want to depend on nothing else connecting to the status API for correctness.

This discussion is pushing me towards suggesting that we *just* implement a status API (since I think we need it any way), and we can implement a real-time API once we have a working status API.

**Comment by meejah** [comment]
> This is only true if no other client has connected between when the service has started and when a given client connects. I don't think we want to depend on nothing else connecting to the status API for correctness.

That is correct. Why shouldn't we depend on that behavior, though? (If you don't want anything else connecting, then it shouldn't gain permission to connect ... right?)

**Comment by meejah** [comment]
> This discussion is pushing me towards suggesting that we just implement a status API (since I think we need it any way), and we can implement a real-time API once we have a working status API.

Hmm, okay .. I think we're in partial agreement: I only want to implement one thing as well. The difference is that I only want to implement the real-time status API and not anything else.

**Comment by tomprince** [comment]
> > This is only true if no other client has connected between when the service has started and when a given client connects. I don't think we want to depend on nothing else connecting to the status API for correctness.
> 
> That is correct. Why shouldn't we depend on that behavior, though? (If you don't want anything else connecting, then it shouldn't gain permission to connect ... right?)

We create localhost http endpoint, with the API key stored on disk. I'm not sure how we can prevent other processes (such as a CLI client) connecting. Unless we do something like provide a way to start a magic-folder service which doesn't store the api key on-disk, or uses something like a pipe that grid-sync connects to, instead of a localhost TCP socket (but both of those come with a host of other complexity that I don't think we want to deal with).

> Hmm, okay .. I think we're in partial agreement: I only want to implement one thing as well. The difference is that I only want to implement the real-time status API and not anything else.

I don't think you can use *just* that and get correct behavior. You can probably get behavior that  is correct most of the time, but will give no indication when it doesn't. With a a status API and polling, you get correct behavior, at a bit of a performance and responsiveness cost.

**Comment by meejah** [comment]
I guess I've been thinking of this as a stream of events and concentrating on the immediate case of "one other app controls magic-folder" ...

...anyway, so with a polling API if we actually wanted a timeline of events you'd just get the entire set of events at the "status" endpoint, approximately, right? (It seems like you're thinking of this as there is some state that exists, and updates to it are published via this WebSocket API?)

So, I suppose it depends what we really want to communicate here longer term? (For the current, single event either works of course .. unless we added a timestamp to when syncing started and stopped). We do know that we want some sort of "progress" in the immediate-term, too.

For example, if we added "more detail" we might add an event like "file 'foo.mp4' started syncing at X" and "file 'foo.mp4' completed syncing at Y" events ... or for the "state" type approach we could encode the same information as a map of "name" -> metadata with metadata containing "up to date / syncing" (and the start/stop times if it synced while the daemon was running). Either way should allow gridsync (or whatever) to display similar information (e.g. a UI could *display* a timeline, or a "history" with either shape of data).

If we conclude we want "a timeline of events", we could just always keep all events (since we started) and send them to any new client (thus giving all clients the same view). This would be comparable to a polling API which returned the whole timeline.

If instead we conclude we want "a single state, that updates" we could still use a WebSocket endpoint similar to #340 but with a slight change: as soon as any client connects, it gets a "here is the current state" message. After that it just gets update messages (for right now, these could just be the same since the state is small).

Both ways eliminate the need for the WebSocket API to be optional or to have different behaviors for different clients (or a different behavior without any client connected).

I think after writing all that out, I prefer the "state" approach .. it might make certain things slightly harder to represent but probably results in more-consistent UIs.

This means: any client that connects to the WebSocket endpoint receives an initial message that encodes "the entire state" and subsequent messages that might be "here's the entire state" (again) or might be some "update" message (TBD in future). For now we'd only ever send "here is the entire state" messages. (If GridSync wanted to show "how long has syncing being going on" the state could include a timestamp when it started..)
